File: tools/polyglot/NeosAudit.kt
package neos.polyglot

import java.io.File

fun main(args: Array<String>) {
    val rootPath = if (args.isNotEmpty()) args[0] else "."
    println("\u001B[1;36m[Kotlin::Audit]\u001B[0m Checking NeOS profile structure at: $rootPath")

    val requiredFiles = listOf(
        "profile/profiledef.sh",
        "profile/pacman.conf",
        "profile/grub/grub.cfg",
        "profile/syslinux/syslinux.cfg",
        "profile/packages.x86_64",
        "profile/airootfs/etc/pacman.d/neos-mirrorlist"
    )

    val missing = requiredFiles.filter { !File(rootPath, it).exists() }
    if (missing.isNotEmpty()) {
        System.err.println("❌ Missing required profile files: ${missing.joinToString(", ")}")
        System.exit(1)
    }

    val pkgFile = File(rootPath, "profile/packages.x86_64")
    val packages = pkgFile.readLines()
        .map { it.trim() }
        .filter { it.isNotEmpty() && !it.startsWith("#") }

    val duplicates = packages.groupingBy { it }.eachCount().filter { it.value > 1 }.keys
    if (duplicates.isNotEmpty()) {
        System.err.println("❌ Duplicate packages found: ${duplicates.joinToString(", ")}")
        System.exit(1)
    }

    println("\u001B[1;32m✓ Kotlin Audit Passed! Verified ${packages.size} unique packages.\u001B[0m")
}

File: tools/polyglot/NeosAudit.swift
import Foundation

let arguments = CommandLine.arguments
let rootPath = arguments.count > 1 ? arguments[1] : "."

print("\u{001B}[1;36m[Swift::Audit]\u{001B}[0m Checking NeOS profile structure at: \(rootPath)")

let fileManager = FileManager.default
let requiredFiles = [
    "profile/profiledef.sh",
    "profile/pacman.conf",
    "profile/grub/grub.cfg",
    "profile/syslinux/syslinux.cfg",
    "profile/packages.x86_64",
    "profile/airootfs/etc/pacman.d/neos-mirrorlist"
]

var missingFiles: [String] = []
for file in requiredFiles {
    let fullPath = (rootPath as NSString).appendingPathComponent(file)
    if !fileManager.fileExists(atPath: fullPath) {
        missingFiles.append(file)
    }
}

guard missingFiles.isEmpty else {
    print("❌ Missing required profile files: \(missingFiles.joined(separator: ", "))")
    exit(1)
}

let pkgPath = (rootPath as NSString).appendingPathComponent("profile/packages.x86_64")
guard let content = try? String(contentsOfFile: pkgPath, encoding: .utf8) else {
    print("❌ Failed to read packages.x86_64")
    exit(1)
}

let lines = content.components(separatedBy: .newlines)
var seen = Set<String>()
var count = 0

for rawLine in lines {
    let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
    if line.isEmpty || line.hasPrefix("#") {
        continue
    }
    if seen.contains(line) {
        print("❌ Duplicate package found in Swift audit: \(line)")
        exit(1)
    }
    seen.insert(line)
    count += 1
}

print("\u{001B}[1;32m✓ Swift Audit Passed! Verified \(count) unique packages.\u{001B}[0m")
